code/obj/machinery/door/airlock.dm

```diff
@@ -28,6 +28,174 @@
 	shock - has a chance of electrocuting its target.
 */
 
+
+/obj/machinery/door/airlock/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "AiAirlock", src.name)
+		ui.open()
+	return TRUE
+
+/obj/machinery/door/airlock/ui_data()
+	var/list/data = list()
+
+	var/list/power = list()
+	power["main"] = secondsMainPowerLost ? 0 : 2 // boolean
+	power["main_timeleft"] = secondsMainPowerLost
+	power["backup"] = secondsBackupPowerLost ? 0 : 2 // boolean
+	power["backup_timeleft"] = secondsBackupPowerLost
+	data["power"] = power
+
+	data["shock"] = secondsElectrified == 0 ? 2 : 0
+	data["shock_timeleft"] = secondsElectrified
+	data["id_scanner"] = !aiDisabledIdScanner
+	data["locked"] = locked // bolted
+	data["welded"] = welded // welded
+	data["opened"] = !density // opened
+
+	var/list/wire = list()
+	wire["main_1"] = !src.isWireCut(AIRLOCK_WIRE_MAIN_POWER1)
+	wire["main_2"] = !src.isWireCut(AIRLOCK_WIRE_MAIN_POWER2)
+	wire["backup_1"] = !src.isWireCut(AIRLOCK_WIRE_BACKUP_POWER1)
+	wire["backup_2"] = !src.isWireCut(AIRLOCK_WIRE_BACKUP_POWER2)
+	wire["shock"] = !src.isWireCut(AIRLOCK_WIRE_ELECTRIFY)
+	wire["id_scanner"] = !src.isWireCut(AIRLOCK_WIRE_IDSCAN)
+	wire["bolts"] = !src.isWireCut(AIRLOCK_WIRE_DOOR_BOLTS)
+	wire["safe"] = !src.isWireCut(AIRLOCK_WIRE_SAFETY)
+
+	data["wires"] = wire
+	return data
+
+/obj/machinery/power/smes/ui_state(mob/user)
+	return tgui_default_state
+
+obj/machinery/door/airlock/ui_status(mob/user, datum/ui_state/state)
+		return min(
+			tgui_silicon_state.can_use_topic(src, user),
+			tgui_broken_state.can_use_topic(src, user),
+			tgui_not_incapacitated_state.can_use_topic(src, user)
+		)
+
+/obj/machinery/door/airlock/ui_act(action, params)
+	if(..())
+		return
+	if(!src.allowed(usr))
+		return
+	if(!src.arePowerSystemsOn())
+		return
```

Comment:
I feel like this is a weird place to check these

I'd expect the UI status to change, or the UI to have never opened in the first place if I couldn't access a door either due to lack or loss of access. Just blocking the users inputs leaves it unclear as to why they can't do anything

Reply:
In addition, perhaps more clear visual feedback for when you've lost/lack access? Just an idea though.

Reply:
I already disable the buttons when the PowerSystems aren't on this only checks if secondsMainPowerLost and 
secondsBackupPowerLost are 0 if the airlock loses power externally the eye will turn red and i'll have the UI update to show there's no external power.